moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.kt

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import com.squareup.moshi.internal.jsonAnnotations
+import com.squareup.moshi.internal.jsonName
+import com.squareup.moshi.internal.knownNotNull
+import com.squareup.moshi.internal.missingProperty
+import com.squareup.moshi.internal.resolve
+import com.squareup.moshi.internal.rethrowCause
+import java.lang.invoke.MethodHandle
+import java.lang.invoke.MethodHandles
+import java.lang.invoke.MethodType.methodType
+import java.lang.reflect.InvocationTargetException
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.RecordComponent
+import java.lang.reflect.Type
+
+/**
+ * A [JsonAdapter] that supports Java `record` classes via reflection.
+ *
+ * **NOTE:** Java records require JDK 16 or higher.
+ */
+internal class RecordJsonAdapter<T>(
+  private val constructor: MethodHandle,
+  private val targetClass: String,
+  componentBindings: Map<String, ComponentBinding<Any>>
+) : JsonAdapter<T>() {
+
+  @JvmRecord
```

Comment:
It was a record before, but doesn't truuuuuly matter here since we have access to a data class. Did appreciate being able to use a JvmRecord though, but can remove since it's ultimately irrelevant

Reply:
```suggestion
```